internal/controller/compute/statefulserverset/datavolume_controller.go

```diff
@@ -0,0 +1,165 @@
+		return v1alpha1.Volume{}, err
+	}
+	// get the volume again before returning to have the id populated
+	kubeVolume, err := k.Get(ctx, name, cr.Namespace)
+	if err != nil {
+		return v1alpha1.Volume{}, err
+	}
+	k.log.Info("Finished creating Data Volume", "name", name)
+
+	return *kubeVolume, nil
+}
+
+// Get - returns a volume kubernetes object
+func (k *kubeDataVolumeController) Get(ctx context.Context, volumeName, ns string) (*v1alpha1.Volume, error) {
+	obj := &v1alpha1.Volume{}
+	err := k.kube.Get(ctx, types.NamespacedName{
+		Namespace: ns,
+		Name:      volumeName,
+	}, obj)
+	return obj, err
+}
+
+// Delete - deletes the datavolume k8s client and waits until it is deleted
+func (k *kubeDataVolumeController) Delete(ctx context.Context, name, namespace string) error {
+	condemnedVolume, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		return err
+	}
+	if err := k.kube.Delete(ctx, condemnedVolume); err != nil {
+		return fmt.Errorf("error deleting data volume %w", err)
+	}
+	return serverset.WaitForKubeResource(ctx, serverset.ResourceReadyTimeout, k.isDataVolumeDeleted, condemnedVolume.Name, namespace)
+}
+
+// isAvailable - checks if a volume is available
+func (k *kubeDataVolumeController) isAvailable(ctx context.Context, name, namespace string) (bool, error) {
+	obj, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	if obj != nil && obj.Status.AtProvider.VolumeID != "" && strings.EqualFold(obj.Status.AtProvider.State, ionoscloud.Available) {
+		return true, nil
+	}
+	return false, err
+}
+
+func (k *kubeDataVolumeController) isDataVolumeDeleted(ctx context.Context, name, namespace string) (bool, error) {
+	k.log.Info("Checking if data volume is deleted", "name", name, "namespace", namespace)
+	obj := &v1alpha1.Volume{}
+	err := k.kube.Get(ctx, types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}, obj)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			k.log.Info("Data volume has been deleted", "name", name, "namespace", namespace)
+			return true, nil
+		}
+		return false, nil
+	}
+	return false, nil
+}
+
+func fromStatefulServerSetToVolume(cr *v1alpha1.StatefulServerSet, name string, replicaIndex, volumeIndex int) v1alpha1.Volume {
+	vol := v1alpha1.Volume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: cr.Namespace,
+			Labels: map[string]string{
+				statefulServerSetLabel: cr.Name,
+				// todo replace with function
+				fmt.Sprintf(volumeselector.VolumeReplicaIndexLabel, cr.Name+"-"+cr.Spec.ForProvider.Template.Metadata.Name, volumeselector.ResourceDataVolume): fmt.Sprintf("%d", replicaIndex),
+				fmt.Sprintf(volumeselector.VolumeIndexLabel, cr.Name+"-"+cr.Spec.ForProvider.Template.Metadata.Name, volumeselector.ResourceDataVolume):        fmt.Sprintf("%d", volumeIndex),
```

Comment:
Maybe refactor it to something like this:
```go
createVolumeLabel(volumeselector.VolumeReplicaIndexLabel, cr): fmt.Sprintf("%d", replicaIndex),
createVolumeLabel(volumeselector.VolumeIndexLabel, cr):        fmt.Sprintf("%d", volumeIndex),
...
func createVolumeLabel(t string, cr *v1alpha1.StatefulServerSet) string {
	return fmt.Sprintf(t, cr.Name+"-"+cr.Spec.ForProvider.Template.Metadata.Name, volumeselector.ResourceDataVolume)
}
```

Reply:
Will change to this: 
func createVolumeLabelKey(label string, name string) string {
	return fmt.Sprintf(label, name, volumeselector.ResourceDataVolume)
}

Reply:
Instead of this `cr.Name+"-"+cr.Spec.ForProvider.Template.Metadata.Name, volumeselector.ResourceDataVolume)` you can use a function like this:
```
func getParentResourceName(cr *v1alpha1.StatefulServerSet) string {
	return cr.Name + "-" + cr.Spec.ForProvider.Template.Metadata.Name
}
```
You can use in `fromSSSetToVolume` method too

Reply:
done